File: datadriven/python/uq/operations/forcePositivity/operationMakePositive.py
'''
Created on Feb 6, 2015

@author: franzefn
'''
from pysgpp.extensions.datadriven.uq.operations import checkPositivity, \
    insertHierarchicalAncestors, insertPoint, copyGrid, \
    dehierarchize, hierarchize, hasChildren, hasAllChildren
from pysgpp import HashGridPoint, createOperationEval, DataVector, IndexList, \
    createOperationQuadrature, GridType_LinearBoundary, GridType_PolyBoundary
import warnings
from pysgpp.extensions.datadriven.uq.plot.plot2d import plotSG2d
import matplotlib.pyplot as plt
from pysgpp.extensions.datadriven.uq.operations.sparse_grid import getHierarchicalAncestors, \
    insertTruncatedBorder
import numpy as np


class OperationMakePositive(object):

    def __init__(self, grid):
        self.grid = grid
        self.verbose = False

    def setInterpolationAlgorithm(self, algorithm):
        self.algorithm = algorithm

    def makeCurrentNodalValuesPositive(self, grid, alpha):
        nodalValues = dehierarchize(grid, alpha)
        cnt = 0
        for i, yi in enumerate(nodalValues):
            if yi < 0:
                nodalValues[i] = 0
                cnt += 1
        if cnt > 0:
            alpha = hierarchize(grid, nodalValues)
            if self.verbose:
                warnings.warn("negative function values at grid points encountered, this should not happen")

        return alpha

    def findCandidatesSweep1d(self, d, gp, alpha, grid, acc,
                              negativeAncestorFound=False):
        gs = grid.getStorage()
        ix = gs.getSequenceNumber(gp)
        negativeAncestorFound |= alpha[ix] < 0.

        # just leaf nodes are possible candidates
        if hasAllChildren(grid, gp) and negativeAncestorFound:
            acc.append(gp)

        # get left child
        level, index = gp.getLevel(d), gp.getIndex(d)
        gp.getLeftChild(d)
        if gs.isContaining(gp):
            self.findCandidatesSweep1d(d, gp, alpha, grid, acc,
                                       negativeAncestorFound)

        # get right child
        gp.set(d, level, index)
        gp.getRightChild(d)
        if gs.isContaining(gp):
            self.findCandidatesSweep1d(d, gp, alpha, grid, acc,
                                       negativeAncestorFound)
        gp.set(d, level, index)

    def findCandidates(self, grid, alpha):
        gs = grid.getStorage()
        candidates = {}

        # lookup dimension-wise
        for d in xrange(gs.getDimension()):
            # compute starting points by level sum
            anchors = []
            for i in xrange(gs.getSize()):
                accLevel = gs.getPoint(i).getLevel(d)
                if accLevel == 1:
                    anchors.append(i)

            while len(anchors) > 0:
                # get next starting node
                ix = anchors.pop(0)
                gp = gs.getPoint(ix)
                acc = []
                self.findCandidatesSweep1d(d, gp, alpha, grid, acc, False)
                # store candidates
                for gp in acc:
                    ix = gs.getSequenceNumber(gp)
                    if ix not in candidates:
                        candidates[ix] = HashGridPoint(gp)

        return candidates.values()

#     def findCandidates(self, grid, alpha):
#         # 1. create a new grid to work on
#         gs = grid.getStorage()
#         candidates = {}
#
#         # run over all grid points...
#         for i in xrange(gs.getSize()):
#             gp = gs.getPoint(i)
#             # ... and check if they are leaf nodes
#             if not hasAllChildren(grid, gp):
#                 # load all hierarchical ancestors
#                 ancestors = getHierarchicalAncestors(grid, gp)
#                 foundNegativeCoefficient = alpha[i] < 0
#                 j = 0
#                 # if there is at least one hierarchical ancestor that is
#                 # negative, then all full grid points in the support of the
#                 # current node can be negative...
#                 while not foundNegativeCoefficient and j < len(ancestors):
#                     ix = gs.getSequenceNumber(ancestors[j][1])
#                     foundNegativeCoefficient = alpha[ix] < 0
#                     j += 1
#
#                 if foundNegativeCoefficient and i not in candidates:
#                     candidates[i] = gp
#
#         return candidates.values()

    def lookupFullGridPointsRec1d(self, grid, alpha, gp, d, p, opEval, maxLevel, acc):
        gs = grid.getStorage()
        level, index = gp.getLevel(d), gp.getIndex(d)

        # if the function value for the left child
        # is negtive, then add it with all its
        # hierarchical ancestors
        gp.getLeftChild(d)
        gp.getStandardCoordinates(p)
        if opEval.eval(alpha, p) < 0:
            acc.append(HashGridPoint(gp))

        if level + 1 < maxLevel:
            self.lookupFullGridPointsRec1d(grid, alpha, gp, d, p, opEval, maxLevel, acc)

        # if the function value for the right child
        # is negtive, then add it with all its
        # hierarchical ancestors
        gp.set(d, level, index)
        gp.getRightChild(d)
        gp.getStandardCoordinates(p)
        if opEval.eval(alpha, p) < 0:
            acc.append(HashGridPoint(gp))

        # store them for next round
        if level + 1 < maxLevel:
            self.lookupFullGridPointsRec1d(grid, alpha, gp, d, p, opEval, maxLevel, acc)

        # reset the grid point
        gp.set(d, level, index)

    def lookupFullGridPoints(self, grid, alpha, candidates):
        acc = []
        gs = grid.getStorage()
        p = DataVector(gs.getDimension())
        opEval = createOperationEval(grid)
        # TODO: find local max level for adaptively refined grids
        maxLevel = gs.getMaxLevel()
        if grid.getType() in [GridType_LinearBoundary, GridType_PolyBoundary]:
            maxLevel += 1

        alphaVec = DataVector(alpha)
        for gp in candidates:
            for d in xrange(gs.getDimension()):
                if 0 < gp.getLevel(d) < maxLevel:
                    self.lookupFullGridPointsRec1d(grid, alphaVec, gp, d, p,
                                                   opEval, maxLevel, acc)
        return acc

    def addFullGridPoints(self, grid, alpha):
        """
        Add all those full grid points with |accLevel|_1 <= n, where n is the
        maximun level of the sparse grid
        @param grid: Grid sparse grid to be discretized
        @param alpha: numpy array hierarchical coefficients
        """

        # make grid isotropic by adding all missing hierarchical ancestors
        newGridPoints = []
        # 1. locate candidates to be refined
        candidates = self.findCandidates(grid, alpha)
        # 2. do full grid search locally
        gridPoinsToBeAdded = self.lookupFullGridPoints(grid, alpha, candidates)
        # 3. insert them in a new grid
        for gp in gridPoinsToBeAdded:
            newGridPoints += insertPoint(grid, gp)
            newGridPoints += insertHierarchicalAncestors(grid, gp)
            if grid.getType() in [GridType_LinearBoundary, GridType_PolyBoundary]:
                newGridPoints += insertTruncatedBorder(grid, gp)

        # recompute the leaf property and return the result
        grid.getStorage().recalcLeafProperty()
        return newGridPoints

    def coarsening(self, grid, alpha, newGridPoints):
        """
        Removes all unnecessary grid points. A grid point is defined as
        unnecessary if it is a leaf node and its hierarchical coefficient is
        negative. This is applied just for the newly added points.
        @param grid: Grid
        @param alpha: numpy array hierarchical coefficients
        @param newGridPoints: newly added grid points
        """
        gs = grid.getStorage()
        newAlpha = alpha
        notAffectedGridPoints = []
        p = DataVector(gs.getDimension())
        # remove all entirely negative weighted sub-branches in the tree
        # just consider the newly added ones by discretization
        iteration = 1
        while True:
            newGrid = copyGrid(grid)
            notAffectedGridPoints = []
            toBeRemoved = IndexList()
            for gp in newGridPoints:
                ix = gs.getSequenceNumber(gp)
                gp.getStandardCoordinates(p)
                # if the grid point is a leaf and has negative weight
                # we dont need it to make the function positive
                if not hasChildren(grid, gp) and newAlpha[ix] < 0.:
                    toBeRemoved.append(ix)
                else:
                    notAffectedGridPoints.append(gp)

            # remove the identified grid points
            if toBeRemoved.size() > 0:
                # copy the old grid
                newGs = newGrid.getStorage()
                # delete the unneccessary grid points from the new grid
                newGs.deletePoints(toBeRemoved)
                newGs.recalcLeafProperty()
                # reset lists
                newGridPoints = notAffectedGridPoints
                # copy the remaining alpha values
                newAlpha = np.ndarray(newGs.getSize())
                for i in xrange(newGs.getSize()):
                    newAlpha[i] = alpha[gs.getSequenceNumber(newGs.getPoint(i))]

                grid, gs, alpha = newGrid, newGs, newAlpha
                iteration += 1
            else:
                break

        return newGrid, newAlpha, newGridPoints

    def makePositive(self, alpha):
        """
        insert recursively all grid points such that the function is positive
        defined. Interpolate the function values for the new grid points using
        the registered algorithm.
        @param alpha: numpy array hierarchical coefficients
        """
        # make sure that the function is positive at every existing grid point
        grid = self.grid
        alpha = self.makeCurrentNodalValuesPositive(grid, alpha)
        # start adding points
        newGridPoints = []
        iteration = 1
        while True:
            # copy the old grid
            newGrid = copyGrid(grid)
            # add all those grid points which have an ancestor with negative
            # coefficient
            if self.verbose:
                print "-" * 60
                print "%i:" % iteration
                print "adding full grid points"
            addedGridPoints = self.addFullGridPoints(newGrid, alpha)

            if len(addedGridPoints) == 0:
                newAlpha = alpha
                break

            if self.verbose:
                print "learning the new density"

            newGridPoints += addedGridPoints
            # set the function value at the new grid points to zero
            newNodalValues = dehierarchize(grid, alpha)
            newNodalValues = np.append(newNodalValues, newGrid.getSize() - len(newNodalValues))
            newAlpha = np.append(alpha, np.zeros(newGrid.getSize() - len(alpha)))
            # compute now the hierarchical coefficients for the newly
            # added points
            newAlpha = self.algorithm.computeHierarchicalCoefficients(newGrid,
                                                                      newAlpha,
                                                                      addedGridPoints)
            # the function does not have to be positive now -> check it again
            if self.verbose:
                print "force function to be positive"
            # check manually if all nodal points are positive
            newNodalValues = dehierarchize(newGrid, newAlpha)
            # collect all negative function values
            forceToBePositive = []
            newGs = newGrid.getStorage()
            for i in xrange(newGs.getSize()):
                gp = newGs.getPoint(i)
                if newNodalValues[newGs.getSequenceNumber(gp)] < 0.:
                    forceToBePositive.append(gp)

            if len(forceToBePositive) > 0:
                # if not, interpolate the function values for the negative
                # grid points
                if self.verbose:
                    warnings.warn("manually forcing the the function to be positive")
                newAlpha = self.makeCurrentGridPositive(newGrid, newAlpha)
                # newAlpha = SetGridPointsToZero().computeHierarchicalCoefficients(grid, newAlpha, forceToBePositive)
                # newAlpha = InterpolateParents().computeHierarchicalCoefficients(grid, newAlpha, forceToBePositive)

            if newGrid.getSize() == grid.getSize():
                break

            if self.verbose:
                fig = plt.figure()
                plotSG2d(newGrid, newAlpha, show_grid_points=True, show_negative=True)
                plt.title("iteration = %i" % iteration)
                fig.show()

                print "%i + %i = %i for discretization" % (grid.getSize(),
                                                           newGrid.getSize() - grid.getSize(),
                                                           newGrid.getSize(),)

            iteration += 1
            grid, alpha = newGrid, newAlpha

        # coarsening: remove all new grid points with negative surplus
        coarsedGrid, coarsedAlpha, _ = self.coarsening(newGrid, newAlpha, newGridPoints)
        if self.verbose:
            print "%i - %i = %i grid size after coarsening" % (newGrid.getSize(),
                                                               newGrid.getSize() - coarsedGrid.getSize(),
                                                               coarsedGrid.getSize(),)

        # scale the result such that the integral over it is one
#         c = createOperationQuadrature(coarsedGrid).doQuadrature(coarsedAlpha)
#         coarsedAlpha.mult(1. / c)

        # security check for positiveness
        checkPositivity(coarsedGrid, coarsedAlpha)

        return coarsedGrid, coarsedAlpha

File: base/src/sgpp/base/operation/hash/OperationMakePositive.cpp
// Copyright (C) 2008-today The SG++ project
// This file is part of the SG++ project. For conditions of distribution and
// use, please see the copyright notice provided with SG++ or at
// sgpp.sparsegrids.org

#include <sgpp/base/operation/hash/OperationMakePositive.hpp>
#include <sgpp/base/operation/hash/OperationMakePositiveCandidateSetAlgorithm.hpp>
#include <sgpp/base/operation/hash/OperationMakePositiveInterpolationAlgorithm.hpp>

#include <sgpp/base/datatypes/DataMatrix.hpp>
#include <sgpp/base/operation/BaseOpFactory.hpp>

#include <vector>

#ifdef _OPENMP
#include <omp.h>
#endif

namespace sgpp {
namespace base {

OperationMakePositive::OperationMakePositive(
    base::Grid& grid, MakePositiveCandidateSearchAlgorithm candidateSearchAlgorithm,
    MakePositiveInterpolationAlgorithm interpolationAlgorithm, bool verbose)
    : grid(grid), minimumLevelSum(0), maximumLevelSum(0), verbose(verbose) {
  // set range of level sums for candidate search
  base::HashGridStorage& gridStorage = grid.getStorage();
  auto numDims = gridStorage.getDimension();
  auto maxLevel = gridStorage.getMaxLevel();
  minimumLevelSum = numDims;
  maximumLevelSum = maxLevel * numDims;

  // set the candidate search algorithm
  switch (candidateSearchAlgorithm) {
    case MakePositiveCandidateSearchAlgorithm::FullGrid:
      candidateSearch = std::make_shared<base::OperationMakePositiveLoadFullGridCandidates>(grid);
      break;
    case MakePositiveCandidateSearchAlgorithm::Intersections:
      candidateSearch =
          std::make_shared<base::OperationMakePositiveFindIntersectionCandidates>(grid);
      break;
    default:
      candidateSearch =
          std::make_shared<base::OperationMakePositiveFindIntersectionCandidates>(grid);
      break;
  }
  candidateSearch->setVerbose(verbose);

  // set the interpolation algorithm
  switch (interpolationAlgorithm) {
    case MakePositiveInterpolationAlgorithm::SetToZero:
      interpolationMethod = std::make_shared<base::OperationMakePositiveSetToZero>();
      break;
    default:
      interpolationMethod = std::make_shared<base::OperationMakePositiveSetToZero>();
      break;
  }
}

OperationMakePositive::~OperationMakePositive() {}

void OperationMakePositive::copyGrid(base::Grid& grid, base::Grid*& newGrid) {
  // create grid of dimensions d - 1 of the same type
  base::HashGridStorage& gridStorage = grid.getStorage();
  auto numDims = gridStorage.getDimension();

  newGrid = base::Grid::createLinearGrid(numDims).release();
  base::HashGridStorage& newGridStorage = newGrid->getStorage();

  // run through grid g and add points to mg
  base::GridPoint gridPoint(numDims);

  for (size_t i = 0; i < gridStorage.getSize(); i++) {
    newGridStorage.insert(gridStorage.getPoint(i));
  }
  newGridStorage.recalcLeafProperty();
}

void OperationMakePositive::makeCurrentNodalValuesPositive(base::DataVector& alpha, double tol) {
  // dehierarchize the grid
  auto opHier = op_factory::createOperationHierarchisation(grid);
  opHier->doDehierarchisation(alpha);

  for (size_t i = 0; i < alpha.getSize(); ++i) {
    if (alpha[i] < tol) {
      alpha[i] = 0.0;
    }
  }

  // hierarchize the result
  opHier->doHierarchisation(alpha);
}

void OperationMakePositive::forceNewNodalValuesToBePositive(base::Grid& grid,
                                                            base::DataVector& alpha,
                                                            std::vector<size_t>& newGridPoints) {
  base::HashGridStorage& gridStorage = grid.getStorage();
  auto opEval = op_factory::createOperationEval(grid);
  DataVector x(gridStorage.getDimension());
  for (size_t i : newGridPoints) {
    gridStorage.getPoint(i).getStandardCoordinates(x);
    double nodalValue = opEval->eval(alpha, x);
    if (nodalValue < 0.0) {
      alpha[i] -= nodalValue;
    }
  }
}

void OperationMakePositive::extractNonExistingCandidatesByLevelSum(
    std::vector<std::shared_ptr<base::HashGridPoint>>& candidates,
    std::vector<std::shared_ptr<base::HashGridPoint>>& finalCandidates, size_t currentLevelSum) {
  base::HashGridStorage& gridStorage = grid.getStorage();
  for (auto& candidate : candidates) {
    if (!gridStorage.isContaining(*candidate) && candidate->getLevelSum() == currentLevelSum) {
      finalCandidates.push_back(candidate);
    }
  }
}

void OperationMakePositive::addFullGridPoints(
    base::Grid& grid, base::DataVector& alpha,
    std::vector<std::shared_ptr<base::HashGridPoint>>& candidates,
    std::vector<size_t>& addedGridPoints, double tol) {
  base::HashGridStorage& gridStorage = grid.getStorage();
  size_t numDims = gridStorage.getDimension();
  size_t numCandidates = candidates.size();
  DataVector x(numDims);

  // check if the function value at the new candidates is negative
  DataMatrix data(numCandidates, numDims);
  for (size_t i = 0; i < candidates.size(); ++i) {
    candidates[i]->getStandardCoordinates(x);
    data.setRow(i, x);
  }

  DataVector nodalValues(numCandidates);
  op_factory::createOperationMultipleEval(grid, data)->mult(alpha, nodalValues);

  // insert the negative ones to the grid
  size_t cntConsidered = 0;
  for (size_t i = 0; i < candidates.size(); ++i) {
    if (nodalValues[i] < tol) {
      ++cntConsidered;
      gridStorage.insert(*candidates[i], addedGridPoints);
    }
  }

  if (verbose) {
    std::cout << "# considered          : " << cntConsidered << " -> " << addedGridPoints.size()
              << " : # new grid points" << std::endl;
  }

  // recompute the leaf property
  grid.getStorage().recalcLeafProperty();

  // resize the alpha vector and set the new values to zero
  alpha.resizeZero(gridStorage.getSize());
}

void OperationMakePositive::makePositive(base::Grid*& newGrid, base::DataVector& newAlpha,
                                         bool resetGrid) {
  // copy the current grid
  if (resetGrid) {
    copyGrid(grid, newGrid);
  }

  // force the nodal values of the initial grid to be positive
  makeCurrentNodalValuesPositive(newAlpha);
  size_t currentLevelSum = minimumLevelSum;

  std::vector<size_t> addedGridPoints;
  std::vector<std::shared_ptr<HashGridPoint>> candidates;

  while (currentLevelSum <= maximumLevelSum) {
    if (verbose) {
      std::cout << "current level sum     : " << currentLevelSum << "/" << maximumLevelSum
                << std::endl;
    }

    // search the next candidate set
    candidateSearch->nextCandidates(*newGrid, newAlpha, currentLevelSum, candidates);

    if (verbose) {
      std::cout << "# candidates found    : " << candidates.size();
    }

    if (candidates.size() > 0) {
      // remove existing candidates from the candidate set and sort the rest by level sum
      std::vector<std::shared_ptr<base::HashGridPoint>> finalCandidates;
      extractNonExistingCandidatesByLevelSum(candidates, finalCandidates, currentLevelSum);

      if (verbose) {
        std::cout << " >= " << finalCandidates.size() << std::endl;
      }

      // add the those candidates for which the sparse grid function is negative
      addFullGridPoints(*newGrid, newAlpha, finalCandidates, addedGridPoints);

      if (verbose) {
        std::cout << "# new grid points     : " << addedGridPoints.size() << " -> "
                  << newGrid->getSize() << std::endl;
      }

      // update the hierarchical coefficients for the new points
      interpolationMethod->computeHierarchicalCoefficients(*newGrid, newAlpha, addedGridPoints);
    } else {
      if (verbose) {
        std::cout << std::endl;
      }
    }

    if (verbose) {
      std::cout << "--------------------------------------------------------" << std::endl;
    }

    // increment the level sum
    ++currentLevelSum;
    addedGridPoints.clear();
  }
}

} /* namespace base */
} /* namespace sgpp */
